scio-smb/src/main/java/org/apache/beam/sdk/extensions/smb/SortedBucketSink.java

```diff
@@ -790,9 +791,12 @@ public final WriteResult expand(PCollection<KV<K, V>> input) {
                     new DoFn<KV<K, V>, Void>() {
                       @ProcessElement
                       public void processElement(ProcessContext c) {
-                        if (!bucketMetadata
-                            .extractKey(c.element().getValue())
-                            .equals(c.element().getKey())) {
+                        final K key = bucketMetadata.extractKey(c.element().getValue());
+                        final boolean keysMatch =
+                            key == null
+                                ? c.element().getKey() == null
+                                : key.equals(c.element().getKey());
```

Comment:
This reminds me, for Avro, we're relying on `exrtraKey` returns `String` instead of `CharSequence`? Doesn't seem the case looking at `AvroBucketMetadata`. Do we want to compare encoded bytes or deep compare?

Reply:
ah yeah that's a good point. I can update to compare bytes, that might be simplest.